### PR TITLE
nghttp2 => 1.24.0 + --enable-app

### PR DESCRIPTION
**Pull request by ixmatus**
This change updates nghttp2 to the latest stable released version, `1.24.0`. Additionally, a boolean argument, `enableApp`, is added to build nghttp2's applications (e.g: `nghttpx`, `nghttpd`, `h2load`, etc.)

###### Motivation for this change

The motivation for this change is two-fold:

1. Pick up the latest stable code for `nghttp2`.
2. Be able to build the apps provided by `nghttp2`, which is a prerequisite for a follow-on PR adding a new NixOS module for `nghttpx`.

###### Things done
- [ ] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [x] Linux
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by makefu** [comment]
The PR Looks good but i recommend to squash the three commits into one which introduces the new feature and bumps the version number or rewrite the commits into:
1. Version bump "nghttp2: 1.20.0 -> 1.24.0"
2. Introduction of the new feature "nghttp2: add enableApp"

The commit messages should be in accordance with the CONTRIBUTING document.
I am looking forward for the nghttp2 module PR 👍

**Comment by ixmatus** [comment]
Oh, technically my commit message still doesn't meet `CONTRIBUTING.md`, I'll fix again.

**Comment by makefu** [comment]
compiles and tested binaries, new 'app' related tools seem to be working.

@Mic92 would you be so kind? 👍

**Comment by ixmatus** [comment]
Thanks for the verification @makefu. Is it still the case that the Travis build failures can be ignored like it was a few months ago? If not, I will go investigate...

Otherwise, if we could get this merged that'd be nice so I can put up my PR for the `nghttpx` NixOS module.

**Comment by makefu** [comment]
Hi @ixmatus , you are right - a failing Travis bot has lost its meaning some time ago.

@qknight can you approve this PR?

Thanks!

File: pkgs/development/libraries/nghttp2/default.nix [hunk 1]
```diff
@@ -6,6 +6,7 @@
 , enableAsioLib ? false, boost ? null
 , enableGetAssets ? false, libxml2 ? null
 , enableJemalloc ? false, jemalloc ? null
+, enableApp ? false
```
Comment (fpletz) [inline pkgs/development/libraries/nghttp2/default.nix]: You could alternatively put the apps into their own output. Adding bin to outputs might suffice.

Comment (ixmatus) [inline pkgs/development/libraries/nghttp2/default.nix]: I was trying to follow the derivation as it was as closely as possible but I'll admit I hadn't thought of your suggestion :smile:.

I'll follow your suggestion and push a revision to the commit this evening. Thanks @fpletz.

**Comment by ixmatus** [comment]
@fpletz I've addressed your feedback.

**Comment by ixmatus** [comment]
@globin I'm confused. I tested it and all of the application binaries were installed into the bin output. Is that what you're talking about?

**Comment by globin** [comment]
No this ensures that you can do e.g `${drvWithBinOutput}/bin/executable` instead of having to add `.bin` :)
Sry that last comment was not clear at all..

**Comment by Mic92** [comment]
I think you can also retarget staging for this one. nghttp2 -> curl -> quite a lot of packages.

**Comment by ixmatus** [comment]
@globin oh I see, I'll amend the commit with that, then.

@Mic92 I do not understand what you mean by retargeting for staging.

**Comment by Mic92** [comment]
Change the target branch to `staging` instead of `master`.

**Comment by ixmatus** [comment]
@Mic92 oh, I understand. Okay.

**Comment by ixmatus** [comment]
Don't merge yet. I have one more change to make to fix an issue I just noticed with `nghttpx` in our own test (ocsp cert stapling needs python, so I need to wrap `nghttpx`).

**Comment by ixmatus** [comment]
@fpletz @globin @Mic92 I'd love your feedback on a1c6bf2. I didn't squash it with the other commit because I'm not sure about it.

It fixes an issue I noticed with nghttpx but this means pulling in python for `nghttp2` which I don't think is bad, per-se, but it increases its "size". So I kept the commit separate in-case any of you have a really good reason for me to rethink how I'm solving this...

File: pkgs/development/libraries/nghttp2/default.nix [hunk 2]
```diff
@@ -17,30 +17,40 @@ let inherit (stdenv.lib) optional; in
...
-    sha256 = "fb29d0500b194f11680203aed21aafab241063ec1397cc51ab5cc0957341141b";
+    sha256 = "18ys6p39yvm9wjjzhhlw35c9m8f5gf4dk9jbshibj19q4js1pnv9";
   };
 
-  outputs = [ "out" "dev" "lib" ];
+  outputs = [ "bin" "out" "dev" "lib" ];
 
   nativeBuildInputs = [ pkgconfig ];
-  buildInputs = [ openssl libev zlib c-ares ]
+  buildInputs = [ openssl libev zlib c-ares makeWrapper ]
```
Comment (fpletz) [inline pkgs/development/libraries/nghttp2/default.nix]: makeWrapper is a nativeBuildInput.

**Comment by vcunat** [comment]
Well, it would certainly be better to not to have build-time dependency `fetchurl` -> ... -> `python`.

Perhaps separate the wrapping into another derivation and pass only the unwrapped one to `curl` (as it uses only the libs anyway).

File: pkgs/development/libraries/nghttp2/default.nix (hunk 1 above)
Comment (vcunat) [reply pkgs/development/libraries/nghttp2/default.nix]: Actually, this made $out basically empty, as libs are in $lib, so it's maybe a little "oversplit".  ($out isn't normally referenced)

**Comment by ixmatus** [comment]
@vcunat thanks for taking a look. I'll try to address your concern this evening or tomorrow when I have some spare time.

**Comment by vcunat** [comment]
I merged the first commit to staging now, as it seems safe and there are many bugfixes within; we can improve later.

**Comment by ixmatus** [comment]
@vcunat I'll work on the improvements in b0a28f5 when I have some free time, I appreciate you taking the time to give me feedback on that commit and also merging something of the improvement for upstream :smile:

**Comment by c0bw3b** [comment]
@ixmatus Actually we were already building the apps even when it wasn't wanted. See the builds history of `libnghttp2` (ie nghttp2.lib) on Hydra. The _--enable-apps_ configure option is automatic as long as openssl, libev, zlib and c-ares are in scope.

Moreover there is a confusion between nghttp2 the C library and nghttp2 apps suite. Trying to build both from one multi-outputs derivation just nurture that confusion.
Examples right in the nixpkgs tree are curl and apache : they use the full nghttp2 derivation as build inputs when they really need the lib alone... As a result, one will pull all nghttp2 apps when installing `curlFull` or `apache`. Which is an issue.

I agree with @vcunat that two different derivations would be much better.
We would need 
1. a stable `libnghttp2` that curl (and thus `fetchurl`) or apache could rely upon ;
2. a separate `nghttp2` derivation for {h2load, nghttp, nghttpd, nghttpx} and we would be able to scope systemd or python or anything into this derivation, opening the road for a nghttpx service module.

FWIW Debian/Ubuntu/Arch/FreeBSD provides lib and apps as separate packages.

I'm willing to work on that and come back with a proper PR in time for 18.03.

**Comment by vcunat** [comment]
`curlFull` pulls *only* `nghttp2.lib` for runtime dependencies, but I suppose you meant build-time dependencies.

**Comment by c0bw3b** [comment]
Yes and no, I guess? 😼 
It is just a build-time dependency in the `curl` derivation but :

```bash
[demo@nixos:~]$ nixos-version 
17.09.1812.36a4dc392a (Hummingbird)

[demo@nixos:~]$ find /nix/store/ -name "nghttpx"

[demo@nixos:~]$ nix-env -iA nixos.curlFull
installing ‘curl-7.56.0’
these paths will be fetched (2.80 MiB download, 5.24 MiB unpacked):
  /nix/store/3lg7jw4g54cy7l65bjk5qzwlb5m7cdxy-c-ares-1.13.0
  /nix/store/4z1n838nkpy42ns4ks749cjldgs5z4l2-openldap-2.4.45-dev
  /nix/store/a3ls83adr1dndnqm1r8d1rj50kq5rhl0-libev-4.24
  /nix/store/chkpwifxfmbkm4sdky2kdlhi2wb7ndld-libssh2-1.8.0-dev
>>/nix/store/dq226ai0lcr5kcp7d978q3ma71l6972i-nghttp2-1.20.0
  /nix/store/hsj75yi8k7m427xdg9xbb98qwc8xr92r-curl-7.56.0-devdoc
  /nix/store/jfd7rlwrkahyql7caknnnxpr1mcrfx3s-curl-7.56.0-dev
  /nix/store/kfiy6my66r0vl1kklwnb0q1x93mn04xq-nghttp2-1.20.0-dev
  /nix/store/lmm7dxr0dip3c545xn4gadb9h4jvij0m-curl-7.56.0-man
  /nix/store/pdlgahwrfx627ksx370difq2144y2wnz-curl-7.56.0-debug
  /nix/store/phhy9mspjwgzm4nv0c5c1j0vkpp9zapn-libidn-1.33-bin
  /nix/store/x03p60i19375ir9hsk6ac0wk1jiza30p-libidn-1.33-dev
[...]

[demo@nixos:~]$ tree /nix/store/dq226ai0lcr5kcp7d978q3ma71l6972i-nghttp2-1.20.0
/nix/store/dq226ai0lcr5kcp7d978q3ma71l6972i-nghttp2-1.20.0
├── bin
│   ├── h2load
│   ├── nghttp
│   ├── nghttpd
│   └── nghttpx
└── share
    ├── doc
    │   └── nghttp2
    │       └── README.rst
    ├── man
    │   └── man1
    │       ├── h2load.1.gz
    │       ├── nghttp.1.gz
    │       ├── nghttpd.1.gz
    │       └── nghttpx.1.gz
    └── nghttp2
        └── fetch-ocsp-response

7 directories, 10 files

[demo@nixos:~]$ find /nix/store/ -name "nghttpx"
/nix/store/dq226ai0lcr5kcp7d978q3ma71l6972i-nghttp2-1.20.0/bin/nghttpx
```

**Comment by ixmatus** [comment]
I agree, at this point I think separating the derivations will make things clearer. I still have yet to contribute my nghttpx nixos module so when I have some free time to work on that, if no one has tackled this, I will.

**Comment by vcunat** [comment]
That's just `curlFull.dev` depending on it due to putting it to `propagatedBuildInputs`.

**Comment by matthewbauer** [comment]
@GrahamCOfBorg build nghttp2

**Review by matthewbauer — CHANGES_REQUESTED** [review]
See infinite recursion from @GrahamcOfBorg

**Comment by ixmatus** [comment]
I think we got what we wanted out of this. A few improvements were made (and merged) in subsequent commits too. I will close this.

**Pull request closed without merging**